crates/router/src/core/payment_methods/cards.rs

```diff
@@ -1876,18 +1877,38 @@ pub async fn list_payment_methods(
     .await?;
 
     // filter out connectors based on the business country
-    let filtered_mcas = helpers::filter_mca_based_on_business_profile(all_mcas, profile_id);
+    let filtered_mcas = helpers::filter_mca_based_on_business_profile(all_mcas.clone(), profile_id);
 
     logger::debug!(mca_before_filtering=?filtered_mcas);
 
+    let mut builder = cgraph::ConstraintGraphBuilder::new();
     let mut response: Vec<ResponsePaymentMethodIntermediate> = vec![];
     for mca in &filtered_mcas {
         let payment_methods = match &mca.payment_methods_enabled {
             Some(pm) => pm.clone(),
             None => continue,
         };
+        if let Err(e) = make_pm_graph(
+            &mut builder,
+            payment_methods,
+            mca.connector_name.clone(),
+            pm_config_mapping,
+            &state.conf.mandates.supported_payment_methods,
+            &state.conf.mandates.update_mandate_supported,
+        ) {
+            logger::error!("Failed to construct constraint graph for list payment methods {e:?}");
+        }
+    }
+
+    let graph = builder.build();
```

Comment:
We need to cache the constraint graph. This will require creating a Moka cache for the constraint graph as well as implementing the cache invalidation logic via redis pub-sub.

@Sarthak1799 has implemented the same for routing so you can take his help.